### PR TITLE
[docs-metrics] Customer ExemplarReservoir and View API configuration

### PR DESCRIPTION
**Pull request by CodeBlanch**
Relates to #2527

## Changes

* Adds documentation which shows how to implement custom `ExemplarReservoir`s and use the View API to configure them for a metric

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)

### CODE REVIEW

File: docs/metrics/extending-the-sdk/README.md [hunk 1]
```diff
@@ -72,12 +71,88 @@ to the `MeterProvider` as shown in the example [here](./Program.cs).
...
+    public HighestValueExemplarReservoir()
+        : base(capacity: 1)
+    {
+    }
+
+    public override void Offer(in ExemplarMeasurement<long> measurement)
+    {
+        if (!this.previousValueLong.HasValue || measurement.Value > this.previousValueLong.Value)
+        {
+            lock (this.lockObject)
```
Comment (cijothomas) [inline docs/metrics/extending-the-sdk/README.md:113]: should we document the need of lock in the docs too?

File: docs/metrics/extending-the-sdk/README.md [hunk 2]
```diff
@@ -72,12 +71,88 @@ to the `MeterProvider` as shown in the example [here](./Program.cs).
...
 
-Not supported.
+> [!NOTE]
+> `ExemplarReservoir` is an experimental API only available in pre-release
+  builds. For details see:
+  [OTEL1004](../../diagnostics/experimental-apis/OTEL1004.md).
 
-## ExemplarReservoir
+Custom [ExemplarReservoir](../customizing-the-sdk/README.md#exemplarreservoir)s
+can be implemented to control how `Exemplar`s are recorded for a metric:
```
Comment (cijothomas) [inline docs/metrics/extending-the-sdk/README.md:82]: perhaps a note here asking users to open an issue with the need of authorizing custom reservoirs, so we'll be able to get a feel of what are the things users are after....

File: docs/metrics/extending-the-sdk/README.md [hunk 3]
```diff
@@ -72,12 +71,88 @@ to the `MeterProvider` as shown in the example [here](./Program.cs).
...
+  control the number of `Exemplar`s which may be recorded by the
+  `ExemplarReservoir`.
+* The `virtual` `OnCollected` method is called after the `ExemplarReservoir`
+  collection operation has completed and may be used to implement cleanup or
+  reset logic.
+* The `bool` `ResetOnCollect` property on `ExemplarReservoir` is set to `true`
+  when delta aggregation temporality is used for the metric using the
+  `ExemplarReservoir`.
+
+```csharp
```
Comment (cijothomas) [inline docs/metrics/extending-the-sdk/README.md:97]: describe a bit about what this example reservoir does.

File: docs/metrics/extending-the-sdk/README.md (hunk 1 above)
Comment (CodeBlanch) [reply docs/metrics/extending-the-sdk/README.md:129]: Added. Also added a note about throwing exceptions.

File: docs/metrics/extending-the-sdk/README.md (hunk 2 above)
Comment (CodeBlanch) [reply docs/metrics/extending-the-sdk/README.md:84]: Added

File: docs/metrics/extending-the-sdk/README.md (hunk 3 above)
Comment (CodeBlanch) [reply docs/metrics/extending-the-sdk/README.md:113]: Added

File: docs/metrics/customizing-the-sdk/README.md [hunk 4]
```diff
@@ -331,6 +331,28 @@ var meterProvider = Sdk.CreateMeterProviderBuilder()
     .Build();
 ```
 
+### Changing the ExemplarReservoir for a Metric
+
+To set the [ExemplarReservoir](#exemplarreservoir) for an individual metric, use
```
[the remark this answers is not in the document]
Comment (CodeBlanch) [reply docs/metrics/customizing-the-sdk/README.md]: I don't feel this is really necessary. The `ExemplarReservoir` link/anchor there jumps down to the section with all that detail.

I did add this though: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5624/commits/9dcb5be15ca29b6ec27c1eb8a5cd5a4d46f0918f

File: docs/metrics/customizing-the-sdk/README.md [hunk 5]
```diff
@@ -251,6 +251,32 @@ by using Views.
...
+To set the [ExemplarReservoir](#exemplarreservoir) for an instrument, use the
+`MetricStreamConfiguration.ExemplarReservoirFactory` property on the View API:
+
+> [!NOTE]
+> `MetricStreamConfiguration.ExemplarReservoirFactory` is an experimental API only
+  available in pre-release builds. For details see:
+  [OTEL1004](../../diagnostics/experimental-apis/OTEL1004.md).
+
+> [!IMPORTANT]
+> Setting `MetricStreamConfiguration.ExemplarReservoirFactory` alone will NOT
```
Comment (vishweshbankwar) [inline docs/metrics/customizing-the-sdk/README.md:265]: Could we show this in the example below as well?. A comment can be added about how to change the filter.

Comment (CodeBlanch) [reply docs/metrics/customizing-the-sdk/README.md:265]: In the alert there is a link to the exemplar filter section which has that example. We could mirror some of that content here but I think duplicating content (in general) is bad practice. Also how much to duplicate? If we just show calling `SetExemplarFilter` here users might copy that without realizing `OTEL_DOTNET_EXPERIMENTAL_METRICS_EXEMPLAR_FILTER_HISTOGRAMS` is probably a better mechanism 🤣

Comment (vishweshbankwar) [reply docs/metrics/customizing-the-sdk/README.md:265]: I missed the link :).
The code snippet below looks self-contained as opposed to other snippets above for views. Users might simply copy paste and try running that. May be we can just follow the format used in other view(s) examples?

Comment (CodeBlanch) [reply docs/metrics/customizing-the-sdk/README.md:265]: Good call! Updated

**Review by vishweshbankwar — APPROVED** [review]

**Review by rajkumar-rangaraj — APPROVED** [review]

**Pull request merged**